packages/preferences/src/browser/abstract-resource-preference-provider.ts

```diff
@@ -15,22 +15,23 @@
  ********************************************************************************/
 
 import { inject, injectable, postConstruct } from 'inversify';
-import * as jsoncparser from 'jsonc-parser';
+import { JSONExt } from '@phosphor/coreutils';
+import { Disposable, MaybePromise, MessageService, Resource, ResourceProvider } from '@theia/core';
+import { PreferenceProvider, PreferenceSchemaProvider, PreferenceScope, PreferenceProviderDataChange } from '@theia/core/lib/browser';
 import URI from '@theia/core/lib/common/uri';
-import { Resource, ResourceProvider, MaybePromise, MessageService } from '@theia/core';
-import { PreferenceProvider } from '@theia/core/lib/browser/preferences';
+import * as jsoncparser from 'jsonc-parser';
 
 @injectable()
 export abstract class AbstractResourcePreferenceProvider extends PreferenceProvider {
 
     // tslint:disable-next-line:no-any
     protected preferences: { [key: string]: any } = {};
+    protected resource: Promise<Resource>;
+    protected onDidResourceChanged: Disposable | undefined;
```

Comment:
why do you need this property?

Reply:
workspace preferences are stored
1) in the ".theia/settings.json" if the workspace data is not saved in a workspace file, or
2) in the workspace file under the "settings" property, if a workspace file exists.

when the state transitions from #1 to the #2, WorkspacePreferenceProvider is not interested in changes made to ".theia/settings.json" any more (an instance of FolderPreferenceProvider will take care of it instead), and therefore, i want to dispose `onDidResourceChanged` when the state transition happens.